custom/inddex/food.py

```diff
@@ -188,67 +228,81 @@ def age_range(self):
-            if val is None:
-                return MISSING
-            return val
-
-        return [_format(getattr(self, column.slug)) for column in INDICATORS]
-
-
-class FoodCaseRow(FoodRow):
-    """A food item directly corresponding to a case in the UCR"""
-
-    def __init__(self, ucr_row, fixtures):
-        super().__init__(ucr_row['food_code'], ucr_row, fixtures)
-        self.caseid = ucr_row['doc_id']
+    @property
+    def recipe_id(self):
+        if self.is_recipe:
+            return self.caseid
+        return self.recipe_case_id or self.already_reported_recipe_case_id or 'NO_RECIPE'
 
     def __getattr__(self, name):
-        # If it's an indicator in the UCR that hasn't been explicitly set, return that val
         if name in _INDICATORS_BY_SLUG:
-            indicator = _INDICATORS_BY_SLUG[name]
-            return self.ucr_row[indicator.slug] if indicator.in_ucr else None
-        return super().__getattr__(name)
-
-
-class RecipeIngredientRow(FoodRow):
-    """A food item inferred from a recipe"""
-    is_ingredient = "yes"
-
-    def __init__(self, ucr_row, fixtures, ingredient):
-        # ucr_row is data from the parent food case
-        # ingredient is static info for this ingredient from the recipes fixture
-        super().__init__(ingredient.ingr_code, ucr_row, fixtures)
+            if self._is_std_recipe_ingredient:
+                # If it's an indicator that hasn't been explicitly set, check if it can
+                # be pulled from the food fixture or from the parent food case's UCR
+                indicator = _INDICATORS_BY_SLUG[name]
+                if indicator.in_food_fixture:
+                    return getattr(self.fixtures.foods[self.food_code], indicator.slug)
+                if indicator.is_recall_meta:
+                    return self.ucr_row[indicator.slug]
+                return None
+            else:
+                # If it's an indicator in the UCR that hasn't been explicitly set, return that val
+                indicator = _INDICATORS_BY_SLUG[name]
+                return self.ucr_row[indicator.slug] if indicator.in_ucr else None
+
+        raise AttributeError(f"FoodRow has no definition for {name}")
+
+
+class RecipeRowGenerator:
+    """Contains all rows of a given recipe. Handles calculations outside the scope of a single row"""
+
+    def __init__(self, recipe_id, rows):
+        self._all_rows = rows
+
+        recipe_possibilities = [row for row in rows if row.is_recipe]
+        self._recipe_row = recipe_possibilities[0] if len(recipe_possibilities) == 1 else None
+        self._recipe_ingredients = [row for row in rows if not row.is_recipe]
+
+    def iter_rows(self):
+        if not self._recipe_row:
+            yield from _yield_non_recipe_rows(self._all_rows)
+
+        for row in [self._recipe_row] + self._recipe_ingredients:
+            yield [_format(self._get_val(row, column.slug)) for column in INDICATORS]
+
+    def _get_val(self, row, slug):
```

Comment:
@kaapstorm What do you think of this method?  I need to override these few indicators, so I just did an if/else list, rather than futz with an overriden `__getattr__`, as I think that's generally best avoided.  Still, I'm using that pattern in the `FoodRow` class, so I could do the same thing here for consistency - add a tag to the indicators used here and look for a class attribute by the same name.

Reply:
I agree with your choice to override `__getattr__` in `FoodRow`, because you're fetching a value from `self.ucr_row` as a fallback to `self`. I think that's what overriding `__getattr__` is for.

But I don't think this is the same thing. I wouldn't override `__getattr__` here. I think this method is simpler and more readable than that would be.

Are you thinking of something like this ...
```python
    def _get_val(self, row, slug):
        if (
            slug in _INDICATORS_BY_SLUG
            and _INDICATORS_BY_SLUG[slug].is_non_row_indicator  # or something
            and getattr(self, 'check_' + slug)(row)
        ):
            return getattr(self, 'get_' + slug)(row)
        return getattr(row, slug)
```
... ? If the number of special slugs increases, having methods for each indicator could be better than having a longer list of `if` clauses.

Reply:
yeah, that's basically what I had in mind, though I was thinking:
```python
    def _get_val(self, row, slug):
        if (
            slug in _INDICATORS_BY_SLUG
            and _INDICATORS_BY_SLUG[slug].is_non_row_indicator  # or something
        ):
            return getattr(self, slug)(row)
        return getattr(row, slug)
```
That is, `is_non_row_indicator == True` should be enough of an assurance that there's a method of the same name on this class.  I'd also prefer to avoid dynamism on the method name so as to not break grep.

I suppose the point is moot now though, since the next PR switched this stuff back to the `FoodRow` class.